### ISSUE TITLE
Packets NATed to TailScale peer silently disappear

### ISSUE THREAD

**Issue by frantisekhanzlikbl** [tailscale/tailscale#8692]
### What is the issue?

If a tailnet device uses iptables/nftables to DNAT packets from outside of the tailnet to another device on the tailnet, the packets get to the `tailscale0` interface but then disappear without reaching the destination device.
	
To put it more concretely:

I have a WireGuard network setup with two peers:
* A, which serves as a public gateway for services on other devices
* B, which is running a service that should be reachable though some port on A
On host A, I had a nftables chain which would DNAT packets from the Internet to B over WireGuard.
```nftables
chain prerouting {
	type nat hook prerouting priority -100;
	tcp dport 25565 dnat ip to BBB.BBB.BBB.BBB:25565;
}
```
`BBB.BBB.BBB.BBB` being B's tailnet address.
On host B, further firewall rules would track all packets coming from the WG interface and automatically use host A as a gateway for replies to them.

This setup worked well, but over time the need arose to use the network even with clients behind NAT, so naturally I reached for TailScale as the solution. Unfortunately, after migrating the network to TS, the DNAT on host A stopped working.

To be more precise: using tcpdump shows the packets arriving to the host A and being correctly sent to the TS interface
```console
# tcpdump -qni any port 25565
20:46:43.484805 enp1s0 In  IP xxx.xxx.xxx.xxx.46442 > aaa.aaa.aaa.aaa.25565: tcp 0
20:46:43.484862 tailscale0 Out IP xxx.xxx.xxx.xxx.46442 > BBB.BBB.BBB.BBB.25565: tcp 0
```
`xxx.xxx.xxx.xxx` being a random device's address which is sending the packets to A over the Internet, `aaa.aaa.aaa.aaa` being A's public (not TS) address and `BBB.BBB.BBB.BBB` being B's tailnet address.

Tracing them using nftables' tracing shows them flowing through the entire firewall without being dropped anywhere. So far, so good. The first indication of trouble is this line in TS logs on host A:
```console
tailscaled[932]: open-conn-track: timeout opening (TCP xxx.xxx.xxx.xxx:46442 => BBB.BBB.BBB.BBB:25565) to node [bBbBb]; online=yes, lastRecv=5s
```
On host B, the packets never seem to arrive. They don't show up in tcpdump or nftables' tracing monitor.

What does show up on host B from time to time are these weird packets:
```console
# tcpdump -qni any port 25565
21:00:23.599411 lo    In  IP6 ::1.25565 > ::1.43538: tcp 0
21:00:23.599426 lo    In  IP6 ::1.43538 > ::1.25565: tcp 0
```
Unfortunately, I'm not really sure whether these are in any way related to the problem, as they seem to pop up rather randomly.

Interestingly, sending the testing packets from the same testing host but through the tailnet too, everything seems to work as expected. The packet gets forwarded from A to B, B sends a reply to A and the reply arrives to C. Seemingly the only interesting difference is that the source IP of the packet is on the tailnet too.

### Steps to reproduce

1. Create a simple tailnet with two devices: A and B
2. Setup a DNAT on A forwarding packets to B over TS.
3. Send some testing packets from the internet to A
4. Observe that the packets never make it through the tailnet to host B.

### OS

Linux

### OS version

NixOS 23.11 (Tapir)

### Tailscale version

1.44.0

### Other software

Pretty normal NixOS setup on both hosts without much weirdness: Firewall, RP filtering (disabling it makes no difference), `net.ipv4.conf.all.forwarding = 1`.
Host A is a Hetzner Cloud machine without any special networking-related features.
Host B is a desktop PC inside a home network.

### Bug report

BUG-4e91957e9cfeb9900c25165686433a4342bca6d59e153107635702d7ae689ac2-20230723192258Z-5712a443b3224618

**Comment by bradfitz** [comment]
If Tailscale doesn't know about that range then, yes, packets from mystery addresses will be dropped.

You'll want to advertise a route from the node that's sending those packets so the others don't reject it.

**Comment by frantisekhanzlikbl** [comment]
Ah, that kinda makes sense. Unfortunately, I'm not really sure how to do the advertising correctly. Given that the gateway might forward packets originating from anywhere on the internet, I tried adding `--advertise-routes=0.0.0.0/0,::/0` to the gateway's `tailscaled` args. That "subnet" however doesn't show up in the machine's route settings in the admin panel. Instead, the node can be allowed to be an exit node. This does make sense, as the functionality is pretty much equal. When I added `--accept-routes` to the other machine's args though, it didn't fix the problem. I assume that this is due to the route not really being considered a route, but I'm not sure how to fix it. I don't want to use the gateway machine as my exist node for all traffic, just for specific packets which I route to it manually. :-/

**Comment by frantisekhanzlikbl** [comment]
For future reference, since I've had to work around this on multiple different occasions, I can achieve the desired effect by:
1) advertising an exit node from A,
2) accepting it from B, and, most importantly,
3) bypassing the routing rules set up by tailscale on host B via the following rule:
    ```
    ip rule add pref 5269 from all lookup main
    ```
4) routing the packets that should use the exit node to the tailscale device. a simple example would be the following:
    ```
    ip rule add pref 5000 from all fwmark 0x{{some fwmark}} lookup {{routing table name}}
    ip route add {{exit node addr}} dev tailscale0
    ip route add default via {{exit node addr}} table {{routing table name}}
    ```

I do not like this at all. It is brittle, opaque, and leads to a complete loss of connection if the bypass rule ever stops working, but it does its job.

**Comment by MatthewCash** [comment]
> For future reference, since I've had to work around this on multiple different occasions, I can achieve the desired effect by:
> 
>     1. advertising an exit node from A,
> 
>     2. accepting it from B, and, most importantly,
> 
>     3. bypassing the routing rules set up by tailscale on host B via the following rule:
>        ```
>        ip rule add pref 5269 from all lookup main
>        ```
> 
>     4. routing the packets that should use the exit node to the tailscale device. a simple example would be the following:
>        ```
>        ip rule add pref 5000 from all fwmark 0x{{some fwmark}} lookup {{routing table name}}
>        ip route add {{exit node addr}} dev tailscale0
>        ip route add default via {{exit node addr}} table {{routing table name}}
>        ```
> 
> 
> I do not like this at all. It is brittle, opaque, can't be used with more than one exit node, and leads to a complete loss of connection if the bypass rule ever stops working, but it does its job. In the future, I am considering setting up an additional wireguard VPN over tailscale, but that feels very unnecessary and unperformant solution for what feels like a simple requirement.

This is exactly what I am doing, it would be nice if tailscale had an option to either add more wireguard AllowedIps or to disable modifying the routing table.

**Comment by frantisekhanzlikbl** [comment]
> This is exactly what I am doing, it would be nice if tailscale had an option to either add more wireguard AllowedIps or to disable modifying the routing table.

Yep. Having a way to forbid Tailscale from modifying the RPDB at all would save me **many** headaches. I'm currently working on yet another problem of a similar nature, and the solution of having a spinning systemd service that attempts to undo everything TS touches is… unsatisfactory. :slightly_smiling_face:
